File: src/Solutions/Euler098.hs
module Solutions.Euler098 where
import Utils.Input(getInput)
import Utils.List(combinations, uniques)
import Data.Char
import qualified Data.IntSet as IntSet
import Data.List
import qualified Data.Map.Strict as Map
import Data.Maybe

type Substitution = Map.Map Char Int
-- maximum length of word in file is 14
squares = IntSet.fromList $ takeWhile (<10^14) $ map (^2) [1..] :: IntSet.IntSet
isSquare :: Int -> Bool
isSquare n = IntSet.member n squares

cmpStr :: String -> String -> Ordering
cmpStr s1 s2
    | length s1 < length s2 = GT
    | length s1 > length s2 = LT
    | length s1 == length s2 = compare s1 s2
grpStr :: String -> String -> Bool
grpStr s1 s2 = length s1 == length s2
sieve :: [String] -> [[String]]
sieve = groupBy grpStr . sortBy cmpStr

singleSubstitute :: Substitution -> Char -> Char
singleSubstitute subst c = intToDigit $ fromJust $ Map.lookup c subst

substitute :: Substitution -> String -> String -> Maybe Int
substitute subst s1 s2 =
    let n1 = read $ map (singleSubstitute subst) s1
        n2 = read $ map (singleSubstitute subst) s2
    in if isSquare n1 && isSquare n2
        then Just $ max n1 n2
        else Nothing

isMappedToZero :: Substitution -> Char -> Bool
isMappedToZero subst c = Map.lookup c subst == Just 0
        
generateSubstitutions :: String -> String -> [Substitution]
generateSubstitutions trailingLetters allLetters = [subst |
    digitsSubset <- combinations (length allLetters) [0..9],
    digitsOrder <- permutations digitsSubset,
    let subst = Map.fromList $ zip allLetters digitsOrder,
    all (not . isMappedToZero subst) trailingLetters]

findSquare :: String -> String -> Maybe Int
findSquare s1 s2
    | sort s1 == sort s2 && length allLetters <= 10
        = maximum [substitute m s1 s2 | m <- generateSubstitutions trailingLetters allLetters]
    | otherwise = Nothing
    where trailingLetters = uniques $ map head [s1, s2] -- can't be substituted to 0
          allLetters = uniques s2

processSingle :: [String] -> [Int]
processSingle l = catMaybes [findSquare s1 s2 | s1 <- l, s2 <- l, s1 < s2]

process :: [[String]] -> Int
process (h:t) =
    let l = processSingle h
    in if null l
        then process t
        else maximum l

parseFile :: String -> [String]
parseFile s = read $ "[" ++ s ++ "]"

euler098 :: IO String
euler098 = show . process . sieve . parseFile <$> getInput 98

File: rust/src/solutions/euler098.rs
use crate::utils::input::get_input;
use crate::utils::numeric::digits::undigits;
use itertools::Itertools;
use std::collections::{HashMap, HashSet};

struct Substitution {
    substitution: HashMap<char, usize>,
}

impl Substitution {
    fn is_mapped_to_zero(&self, c: &char) -> bool {
        self.substitution.get(c).map_or(false, |&v| v == 0)
    }

    fn square_anagrams(&self, squares: &HashSet<usize>, s1: &String, s2: &String) -> Option<usize> {
        let n1 = undigits(
            &s1.chars()
                .map(|c| *self.substitution.get(&c).unwrap())
                .collect(),
        );
        let n2 = undigits(
            &s2.chars()
                .map(|c| *self.substitution.get(&c).unwrap())
                .collect(),
        );
        if squares.contains(&n1) && squares.contains(&n2) {
            Some(std::cmp::max(n1, n2))
        } else {
            None
        }
    }

    fn generate_substitutions(
        trailing_letters: &Vec<char>,
        all_letters: &Vec<char>,
    ) -> Vec<Substitution> {
        (0..10)
            .combinations(all_letters.len())
            .flat_map(|digits_subset| {
                let l = digits_subset.len();
                digits_subset.into_iter().permutations(l)
            })
            .map(|digits_order| Substitution {
                substitution: all_letters
                    .clone()
                    .into_iter()
                    .zip(digits_order.into_iter())
                    .collect(),
            })
            .filter(|substitution| {
                trailing_letters
                    .iter()
                    .all(|c| !substitution.is_mapped_to_zero(c))
            })
            .collect()
    }
}

fn find_square(squares: &HashSet<usize>, s1: &String, s2: &String) -> Option<usize> {
    if s1.len() != s2.len() {
        return None;
    }
    let mut sorted_s1: Vec<char> = s1.chars().collect();
    sorted_s1.sort();
    let mut sorted_s2: Vec<char> = s2.chars().collect();
    sorted_s2.sort();
    let mut all_letters = sorted_s2.clone();
    all_letters.dedup();
    if sorted_s1 == sorted_s2 && all_letters.len() < 10 {
        Substitution::generate_substitutions(
            &vec![s1.chars().next().unwrap(), s2.chars().next().unwrap()],
            &all_letters,
        )
        .into_iter()
        .flat_map(|substitution| substitution.square_anagrams(squares, s1, s2))
        .max()
    } else {
        None
    }
}

pub fn euler098() -> String {
    let squares: HashSet<usize> = (1..10000000).map(|n| n * n).collect();
    let words: Vec<String> = get_input(98)
        .split(',')
        .map(|l| l.trim_matches('"').to_string())
        .collect();
    (0..words.len())
        .flat_map(|i| (i + 1..words.len()).map(move |j| (i, j)))
        .flat_map(|(i, j)| find_square(&squares, &words[i], &words[j]))
        .max()
        .unwrap()
        .to_string()
}
